build/build_utils.py

```diff
@@ -0,0 +1,306 @@
+#  Copyright (C) 2016 - Yevgen Muntyan
+#  Copyright (C) 2016 - Ignacio Casal Quinteiro
+#  Copyright (C) 2016 - Arnavion
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 2 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this program; if not, see <http://www.gnu.org/licenses/>.
+
+# Base building fun / classes, utility fun & var
+
+import os
+import shutil
+import stat
+import sys
+
+global_verbose = False
+global_debug = False
+
+def print_message(msg):
+    print(msg)
+
+def print_log(msg):
+    if global_verbose:
+        print(msg)
+
+def print_debug(msg):
+    if global_debug:
+        print("Debug:", msg)
+
+def error_exit(msg):
+    print("Error:", msg, file=sys.stderr)
+    sys.exit(1)
+
+def convert_to_msys(path):
+    path = path
+    if path[1] != ':':
+        raise Exception('oops')
+    path = '/' + path[0] + path[2:].replace('\\', '/')
+    return path
+
+def _rmtree_error_handler(func, path, exc_info):
+    if not os.access(path, os.W_OK):
+        # Is the error an access error ?
+        os.chmod(path, stat.S_IWUSR)
+        func(path)
+        print_debug('rmtree:read-only file/path (%s)' % (path, ))
+    else:
+        raise
+
+class ordered_set(set):
+    def __init__(self):
+        set.__init__(self)
+        self.__list = list()
+
+    def add(self, o):
+        if not o in self:
+            set.add(self, o)
+            self.__list.append(o)
+
+    def __iter__(self):
+        return self.__list.__iter__()
+
+class Tarball(object):
+    def unpack(self):
+        print_log('Extracting %s to %s' % (self.archive_file, self.builder.working_dir))
+
+        os.makedirs(self.build_dir)
+        self.builder.exec_msys([self.builder.tar, 'ixf', convert_to_msys(self.archive_file), '-C', convert_to_msys(self.build_dir), '' if self.tarbomb else '--strip-components=1'])
+
+        print_log('Extracted %s' % (self.archive_file,))
+
+class MercurialRepo(object):

```

Comment:
Like having a hg_repo.py and git_repo.py or a single file, say build_unpack.py, with tarball, hg, git (and, maybe in the future, zipball or whatever) ?